packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -1485,6 +1504,106 @@ export enum BucketAccessControl {
   AWS_EXEC_READ = 'AwsExecRead',
 }
 
+export interface RoutingRuleCondition {
+  /**
+   * The HTTP error code when the redirect is applied
+   *
+   * In the event of an error, if the error code equals this value, then the specified redirect is applied.
+   *
+   * If both condition properties are specified, both must be true for the redirect to be applied.
+   *
+   * @default - The HTTP error code will not be verified
+   */
+  readonly httpErrorCodeReturnedEquals?: string;
+
+  /**
+   * The object key name prefix when the redirect is applied
+   *
+   * If both condition properties are specified, both must be true for the redirect to be applied.
+   *
+   * @default - The object key name will not be verified
+   */
+  readonly keyPrefixEquals?: string;
+}
+
+export interface ReplaceKeyWith {
+  /**
+   * The specific object key to use in the redirect request
+   *
+   * For example, redirect request to error.html
+   */
+  readonly with: string;
+}
+
+export interface ReplaceKeyPrefixWith {
+  /**
+   * The object key prefix to use in the redirect request
+   */
+  readonly prefixWith: string;
+}
+
+function isReplaceKeyWith(replaceKey?: ReplaceKeyWith | ReplaceKeyPrefixWith): replaceKey is ReplaceKeyWith {
+  return replaceKey !== undefined && replaceKey.hasOwnProperty('with');
+}
+
+function isReplaceKeyPrefixWith(replaceKey?: ReplaceKeyWith | ReplaceKeyPrefixWith): replaceKey is ReplaceKeyPrefixWith {
+  return replaceKey !== undefined && replaceKey.hasOwnProperty('prefixWith');
+}
+
+export interface RoutingRuleProps {
+  /**
+   * The host name to use in the redirect request
+   *
+   * @default - The host name used in the original request.
+   */
+  readonly hostName?: string;
+
+  /**
+   * The HTTP redirect code to use on the response
+   *
+   * Not required if one of the siblings is present.
+   *
+   * @default "301" - Moved Permanently
+   */
+  readonly httpRedirectCode?: string;
+
+  /**
+   * Protocol to use when redirecting requests
+   *
+   * @default - The protocol used in the original request.
+   */
+  readonly protocol?: RedirectProtocol;
+
+  /**
+   * Specifies the object key prefix to use in the redirect request
+   *
+   * Not required if one of the siblings is present.
+   *
+   * @default - The key will not be replaced
+   */
+  readonly replaceKey?: ReplaceKeyWith | ReplaceKeyPrefixWith;
```

Comment:
Unions are not allowed. Use "union-like" classes:

```ts
class ReplaceKey {
  public static with(...);
  public static prefixWith(...);
```

Reply:
Done